src/D2L.CodeStyle.Analyzers/Helpers/ConstantAnalyzer.cs

```diff
@@ -0,0 +1,136 @@
+﻿using System.Collections.Immutable;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Helpers {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class ConstantAnalyzer : DiagnosticAnalyzer {
+		private const string AttributeName = "D2L.CodeStyle.Annotations.Contract.ConstantAttribute";
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+			ImmutableArray.Create(
+				Diagnostics.NonConstantPassedToConstantParameter,
+				Diagnostics.InvalidConstantType
+			);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis( GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics );
+			context.RegisterCompilationStartAction( CompilationStart );
+		}
+
+		public static void CompilationStart(
+			CompilationStartAnalysisContext context
+		) {
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeParameters(
+					ctx,
+					(ParameterSyntax)ctx.Node
+				),
+				SyntaxKind.Parameter
+			);
+
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeArguments(
+					ctx,
+					(ArgumentSyntax)ctx.Node
+				),
+				SyntaxKind.Argument
+			);
+		}
+
+		private static void AnalyzeParameters(
+			SyntaxNodeAnalysisContext context,
+			ParameterSyntax parameterSyntax
+		) {
+			var parameter = context.SemanticModel.GetDeclaredSymbol( parameterSyntax );
+
+			// Parameter is somehow null, so do nothing
+			if( parameter == null ) {
+				return;
+			}
+
+			// Parameter is not [Constant], so do nothing
+			if( !parameter.HasAttribute( context.Compilation, AttributeName ) ) {
+				return;
+			}
+
+			var type = parameter.Type;
+
+			// If the parameter is a valid type, there are no issues
+			if( type.SpecialType != SpecialType.None
+			) {
+				return;
+			}
+
+			// If the parameter is a generic type, there aren't issues yet...
+			// There may be issues when the method is being called, but
+			// those are handled in the argument analyzer
+			if( type.Kind != SymbolKind.NamedType
+			) {
+				return;
+			}

```

Comment:
The terminology of "generic type" here isn't quite write. `IEnumerable<T>` is a generic type. `T` is a type parameter.

```suggestion
			// If the parameter's type is a type parameter then whether
			// it can be constant depends on the type that gets filled and
			// we can't determine that here
			if( type.Kind == SymbolKind.TypeParameter ) {
				return;
			}
```